### PR TITLE
Implement ElixirSense.

### PR DESCRIPTION
**Pull request by PotterDai**
-Implement ElixirSense.
-Add SignatureHelpProvider.
-Upgrade VSCode and Typescript version.

### CODE REVIEW

**Comment by fr1zle** [comment]
Woah, awesome! Gonna check this PR later today. Thank you very much!

File: .vscode/launch.json [hunk 1]
```diff
@@ -21,7 +21,7 @@
 			"args": ["--extensionDevelopmentPath=${workspaceRoot}", "--extensionTestsPath=${workspaceRoot}/out/test" ],
 			"stopOnEntry": false,
 			"sourceMaps": true,
-			"outFiles": ["out/test"],
+            "outFiles": [ "${workspaceRoot}/out/test/**/*.js" ],
 			"preLaunchTask": "npm"
 		}
```
Comment (fr1zle) [inline .vscode/launch.json]: Formatting

File: .vscode/settings.json [hunk 2]
```diff
@@ -5,6 +5,5 @@
 	},
 	"search.exclude": {
 		"out": true // set this to false to include "out" folder in search results
-	},
-	"typescript.tsdk": "./node_modules/typescript/lib" // we want to use the TS server from our node_modules folder to control its version
+	}
 }
```
Comment (fr1zle) [inline .vscode/settings.json]: Any reason for this change?

File: .vscode/tasks.json [hunk 3]
```diff
@@ -23,7 +23,7 @@
 	"args": ["run", "compile", "--loglevel", "silent"],
 
 	// The tsc compiler is started in watching mode
-	"isWatching": true,
+    "isBackground": true,
```
Comment (fr1zle) [inline .vscode/tasks.json]: Formatting?

File: src/elixirSenseDefinitionProvider.ts [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+                            location = new vscode.Location(vscode.Uri.file(filePath), new vscode.Position(0, 0));
+                        }
+                    } else {
+                        location = new vscode.Location(vscode.Uri.file(res), new vscode.Position(0, 0));
+                    }
+                    resolve(location);
+                }
+            });
+            */
+
```
Comment (fr1zle) [inline src/elixirSenseDefinitionProvider.ts]: Maybe delete this?

File: src/elixirSenseHoverProvider.ts [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+            this.server.getDocumentation(document, position, (hover: vscode.Hover) => {
+                if (!token.isCancellationRequested) {
+                    resolve(hover);
+                } else {
+                    console.error('rejecting');
+                    reject();
+                }
+            });
+            */
+
```
Comment (fr1zle) [inline src/elixirSenseHoverProvider.ts]: Maybe delete this?

Comment (PotterDai) [inline src/elixirSenseHoverProvider.ts]: Done! =)

File: src/elixirSenseDefinitionProvider.ts (hunk 4 above)
Comment (PotterDai) [inline src/elixirSenseDefinitionProvider.ts]: Done! =)

**Comment by PotterDai** [comment]
Hi, just did all the follow up cleanings and changes.
The code was done in a bit rush last week so it lacks a bit cleaning, now it should be good.
Please report if there is any errors. Thank you! = )

File: .vscode/settings.json (hunk 2 above)
Comment (PotterDai) [inline .vscode/settings.json]: Hey, this is overwritten by autogenerated file from latest yeoman.
I just added it back. = )

**Comment by fr1zle** [comment]
Okay, the Code is looking fine now ;-)

Hower I am having issues getting this to run. I checked out your changes, inited the submodule and did a clean npm install. I keep getting this error:

```
[vscode-elixir] Cannot decode Erlang term: <<131,116,0,0,0,3,119,5,101,114,114,111,114,119,3,110,105,108,119,7,112,97,121,108,111,97,100,116,0,0,0,3,119,14,97,99,116,117,97,108,95,115,117,98,106,101,99,116,109,0,0,0,8,67,97,104,46,71,97,109,101,119,4,100,111,99,115,116,0,0,0,3,119,9,99,97,108,108,98,97,99,107,115,109,0,0,0,0,119,4,100,111,99,115,109,0,0,0,26,78,111,32,100,111,99,117,109,101,110,116,97,116,105,111,110,32,97,118,97,105,108,97,98,108,101,119,5,116,121,112,101,115,109,0,0,0,0,119,7,115,117,98,106,101,99,116,109,0,0,0,8,67,97,104,46,71,97,109,101,119,10,114,101,113,117,101,115,116,95,105,100,97,6>>
Reason: RangeError: Offset is outside the bounds of the DataView
    at DataView.getUint8 (<anonymous>)
    at _decodeAtom (/home/timm/projects/vscode-elixir/out/src/jem.js:74:39)
    at _decodeObject (/home/timm/projects/vscode-elixir/out/src/jem.js:112:24)
    at _decodeValue (/home/timm/projects/vscode-elixir/out/src/jem.js:55:20)
    at Object.decode (/home/timm/projects/vscode-elixir/out/src/jem.js:18:16)
    at ElixirSenseClient.readPacket (/home/timm/projects/vscode-elixir/out/src/elixirSenseClient.js:58:30)
    at ElixirSenseClient.handleData (/home/timm/projects/vscode-elixir/out/src/elixirSenseClient.js:39:18)
    at emitOne (events.js:96:13)
    at Socket.emit (events.js:188:7)
    at readableAddChunk (_stream_readable.js:176:18)
```

Am I missing anything here?

**Comment by PotterDai** [comment]
Hey, just decoded the Erlang term, the problem seems at jem.ts.
It's weird at my place all projects works fine.
Did you try with another project? Is the result the same?

**Comment by PotterDai** [comment]
Fixed!
The bug is that the code I took from jem.js(https://github.com/inaka/jem.js) ignores atom type 118 and 119. Just fixed here, and I think I should make a pull request there as well = ).

**Comment by fr1zle** [comment]
Yes, that bug seems fixed, great work!

I am going to compare both versions for a bit now, before i merge this.

**Comment by fr1zle** [comment]
This seems to work really well, and I like the added Parameter Help!

Going to merge this now and publish a release later today / this week.

Again, thank you very (!) much for your contribution!

**Comment by PotterDai** [comment]
Sure!
Also just a remind, please add the new configurations to the README file when do the publish.
https://github.com/fr1zle/vscode-elixir/blob/master/package.json#L148
Just in case somebody still prefer the original alchemist-server version, they can switch quite easily.

**Comment by msaraiva** [comment]
@PotterDai @fr1zle Great job! It would be really nice if we could have a centralized repo for the customized `jem.js` and `elixir-sense-client.js`. This way, changes like [this](https://github.com/fr1zle/vscode-elixir/pull/67/commits/7a75847734f4ab8f2fb29a3bc886b7087cc6da2d) could benefit any JS based editor (like Atom). I guess we could have those two files in the `elixir_sense` project. We could then configure a `package.json` so we can publish it as an npm package. What do you think?

BTW. We could also include all needed `.ex` and `.exs` in the package. This way client and server files will always be in sync and we won't have to copy those files manually anymore ;)

**Comment by PotterDai** [comment]
@msaraiva Sure, that's a great idea!
Could you create a new repo for jem.js and elixir-sense-client.js? Then I can make a pull request for the revised version of jem.js .

**Pull request merged**